composer/utils/batch_helpers.py

```diff
@@ -0,0 +1,99 @@
+from typing import Any
+
+from composer.core.types import BatchKeys
+
+
+def batch_get(batch: Any, key: BatchKeys) -> Any:
+    """Indexes into the batch given the key.
+
+    Tries all the common combination of batch and key pairs:
+        * int for sequence
+        * slice for sequence
+        * string for dict
+        * string atribute for object or named_tuple
+        * sequence of ints for sequence
+        * sequence of slices for sequence
+        * sequence of keys for dict
+        * sequence of attribute names for object or named_tuple
+
+    Args:
+        batch (Any): An object that contains the input and label of the items in the batch.
+            Can be any abritrary type that user creates, but we assume some sort of
+            sequence (list, tuple, tensor, array), mapping (dictionary),
+            or attribute store (object with data members, namedtuple).
+        key (BatchKeys): A key to index into the batch. BatchKeys is of type Union[BatchKey Sequence[BatchKey]]
+            and BatchKey is of type Union[int, str, slice].
+
+    Returns:
+        The part of the batch specified by the key. This could be any type depending on
+            what the batch is composed of.
+
+    Raises:
+        RuntimeError if the batch can't be indexed by the key.
+    """
+    try:
+        return batch[key]  # int or slice for sequence or string for dict.
+    except:
```

Comment:
would you consider using explicit guards here instead? e.g.
```
if isinstance(batch, dict):
    return batch[key]
```
etc?

Reply:
I agree with hanlin on this. though we'd probably want to use `Mapping` rather than dict to cover other dictlike stuff.

Reply:
yeah good point. explicit guards makes sense. I wanted to do EAFP instead of LBYL, but I guess in this case LBYL makes more sense. Fixed here:  05fd3f1

Reply:
@ravi-mosaicml and I think that maybe EAFP does make more sense because there might be some custom type that has a __getitem__ function that does not fall under Mapping, Sequence, torch.Tensor, np.ndarray, and it may not be scalable to keep adding more `elif` clauses every time the user breaks the library with a custom type.

Reply:
wdyt, @hanlint, @A-Jacobson ?

Reply:
EAFP could be reasonable here too, but just be sure to catch the right errors (e.g. `batch[key]` could either be a `KeyError` or a `TypeError`, but we should only catch the latter).

Reply:
Ok switched back to EAFP. The code is a little less readable now, but maybe more general to new types. Lemme know what you think, @ravi-mosaicml, @hanlint